src/mlpack/core/data/bow_encoding_impl.hpp

```diff
@@ -0,0 +1,138 @@
+ * @author Jeffin Sam
+ *
+ * Implementation of Bag of Words encoding functions.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_BOW_ENCODING_IMPL_HPP
+#define MLPACK_CORE_DATA_BOW_ENCODING_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "bow_encoding.hpp"
+
+namespace mlpack {
+namespace data {
+
+void Bow::Reset()
+{
+  mappings.clear();
+  originalStrings.clear();
+}
+
+Bow::Bow(const Bow& oldObject)
+{
+  originalStrings = oldObject.originalStrings;
+  std::deque<std::string>::iterator jt = originalStrings.begin();
+  for (auto it = oldObject.originalStrings.begin(); it !=
+      oldObject.originalStrings.end(); it++)
+  {
+    mappings[*jt] = oldObject.mappings.at(*it);
+    jt++;
+  }
+}
+
+void Bow::operator= (const
+    Bow &oldObject)
+{
+  if (this != &oldObject)
+  {
+    mappings.clear();
+    originalStrings.clear();
+    originalStrings = oldObject.originalStrings;
+    std::deque<std::string>::iterator jt = originalStrings.begin();
+    for (auto it = oldObject.originalStrings.begin(); it !=
+        oldObject.originalStrings.end(); it++)
+    {
+      mappings[*jt] = oldObject.mappings.at(*it);
+      jt++;
+    }
+  }
+}
+
+template<typename TokenizerType>
+void Bow::CreateMap(std::string& input,
+    TokenizerType tokenizer)
+{
+  boost::string_view strView(input);
+  boost::string_view token;
+  token = tokenizer(strView);
+  std::size_t curLabel = mappings.size() + 1;
+  while (!token.empty())
+  {
+    if (mappings.find(token) == mappings.end())
+    {
+        originalStrings.push_back(std::string(token));
+        mappings[originalStrings.back()] = curLabel++;
+    }
+    token = tokenizer(strView);
+  }
+}
+
+template<typename MatType, typename TokenizerType>
+void Bow::Encode(const std::vector<std::string>& input,
+                                MatType& output, TokenizerType tokenizer)
+{
+  boost::string_view strView;
+  boost::string_view token;
+  std::vector< std::vector<boost::string_view> > dataset;
+  std::size_t curLabel = mappings.size();
+  for (size_t i = 0; i < input.size(); i++)
+  {
+    strView = input[i];
+    token = tokenizer(strView);
+    dataset.push_back(std::vector<boost::string_view>() );
+    while (!token.empty())
+    {
+      dataset[i].push_back(token);
+      if (mappings.find(token) == mappings.end())
+      {
+        originalStrings.push_back(std::string(token));
+        mappings[originalStrings.back()] = curLabel++;
+      }
+      token = tokenizer(strView);
+    }
+  }
+  output.zeros(dataset.size(), mappings.size());
+  for (size_t i = 0; i < dataset.size(); ++i)
+  {
+    for (size_t j = 0; j < dataset[i].size(); ++j)
+    {
+      output.at(i, mappings.at(dataset[i][j])) = 1;
```

Comment:
Looks like this is the only difference from `DictionaryEncoding` (i.e. this very line). Maybe it's a good idea to merge this classes, how do you think? I mean we can introduce some `EncodingPolicy` which will handle the encoding process.

Reply:
Yes :)